transport/internal/tlsmux/listener.go

```diff
@@ -21,45 +21,130 @@
+type Config struct {
+	Listener  net.Listener
+	TLSConfig *tls.Config
+
+	ServiceName   string
+	TransportName string
+	Meter         *metrics.Scope
+	Logger        *zap.Logger
+}
+
 // listener wraps original net listener and it accepts both TLS and non-TLS connections.
 type listener struct {
 	net.Listener
+
 	tlsConfig *tls.Config
+	observer  *observer
+	logger    *zap.Logger
+
+	serveOnce   sync.Once
+	closeOnce   sync.Once
+	connChan    chan net.Conn
+	stopChan    chan struct{}
+	stoppedChan chan struct{}
 }
 
 // NewListener returns a multiplexed listener which accepts both TLS and
 // plaintext connections.
-func NewListener(lis net.Listener, tlsConfig *tls.Config) net.Listener {
+func NewListener(c Config) net.Listener {
 	return &listener{
-		Listener:  lis,
-		tlsConfig: tlsConfig,
+		Listener:    c.Listener,
+		tlsConfig:   c.TLSConfig,
+		observer:    newObserver(c.Meter, c.Logger, c.ServiceName, c.TransportName),
+		logger:      c.Logger,
+		connChan:    make(chan net.Conn),
+		stoppedChan: make(chan struct{}),
+		stopChan:    make(chan struct{}),
 	}
 }
 
-// Accept returns multiplexed plaintext connection.
+// Accept returns multiplexed plaintext connetion.
+// After close, returned error is errListenerClosed.
 func (l *listener) Accept() (net.Conn, error) {
+	// Starts the connection server only once.
+	l.serveOnce.Do(func() { go l.serve() })
+
+	select {
+	case conn, ok := <-l.connChan:
+		if !ok {
+			return nil, errListenerClosed
+		}
+		return conn, nil
+	case <-l.stopChan:
+		return nil, errListenerClosed
+	}
+}
+
+// Close closes the listener and waits until the connection server drains
+// accepted connections and stops the server.
+func (l *listener) Close() error {
+	var err error
+	l.closeOnce.Do(func() {
+		err = l.Listener.Close()
+		close(l.stopChan)
+		<-l.stoppedChan
+	})
+	return err
+}
+
+// serve starts accepting the connection from the underlying listener and creates a new
+// go routine for each connection for async muxing.
+func (l *listener) serve() {
+	var wg sync.WaitGroup
+
+	defer func() {
+		wg.Wait()
+		close(l.stoppedChan)
+		close(l.connChan)
+	}()
+
 	for {
 		conn, err := l.Listener.Accept()
 		if err != nil {
-			return conn, err
+			return
 		}
 
-		// TODO(jronak): avoid slow connections causing head of the line blocking by spawning
-		// connection processing in separate routine.
+		wg.Add(1)
+		go l.serveConnection(conn, &wg)
+	}
+}
 
-		c, err := l.mux(conn)
-		if err != nil {
-			// Don't return the mux error as caller will shutdown the server on
-			// listener error. Instead, we close the connection and loop around to
-			// acccept the next connection.
-			conn.Close()
-			continue
-		}
-		return c, err
+// serveConnection muxes the given connection and sends muxed connection to the
+// connection channel.
+func (l *listener) serveConnection(conn net.Conn, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	c, err := l.mux(conn)
+	if err != nil {
+		conn.Close()
```

Comment:
Any particular reason to discard the err here without logging/passing?

Reply:
Bad rebase again, the errors were logged and metrics were emitted in the handler. This error is used for closing connection